src/modules/test/Module.c

```diff
@@ -0,0 +1,392 @@
+            info->version.patch = json_object_get_number(object, INFO_VERSION_PATCH);
+            info->version.tweak = json_object_get_number(object, INFO_VERSION_TWEAK);
+
+            if ((lifetime < 0) || (2 < lifetime))
+            {
+                LOG_ERROR("Module info has invalid lifetime: %d", lifetime);
+                status = EINVAL;
+            }
+            else
+            {
+                info->lifetime = lifetime;
+            }
+
+            if ((NULL != licenseUri) && (NULL == (info->licenseUri = strdup(licenseUri))))
+            {
+                LOG_ERROR("Failed to allocate memory for module license URI");
+                status = ENOMEM;
+            }
+
+            if ((NULL != projectUri) && (NULL == (info->projectUri = strdup(projectUri))))
+            {
+                LOG_ERROR("Failed to allocate memory for module project URI");
+                status = ENOMEM;
+            }
+
+            if (NULL == (info->components = (char**)calloc(componentsCount, sizeof(char*))))
+            {
+                LOG_ERROR("Failed to allocate memory for module components");
+                status = ENOMEM;
+            }
+            else
+            {
+                info->componentCount = componentsCount;
+
+                for (int i = 0; i < componentsCount; i++)
+                {
+                    if (NULL == (component = (char*)json_array_get_string(components, i)))
+                    {
+                        LOG_ERROR("Failed to get component name at index: %d", i);
+                        status = EINVAL;
+                        break;
+                    }
+                    else if (NULL == (info->components[i] = strdup(component)))
+                    {
+                        LOG_ERROR("Failed to copy component name at index: %d", i);
+                        status = ENOMEM;
+                        break;
+                    }
+                }
+            }
+
+            if (0 == status)
+            {
+                *moduleInfo = info;
+            }
+            else {
+                FreeModuleInfo(info);
+                info = NULL;
+            }
+
+        }
+    }
+
+    return status;
+}
+
+MANAGEMENT_MODULE* LoadModule(const char* client, const char* bin, const char* name)
+{
+    int status = 0;
+    char* path = NULL;
+    MANAGEMENT_MODULE* module = NULL;
+    MODULE_INFO* info = NULL;
+    size_t pathLength = 0;
+    JSON_Value* value = NULL;
+    MMI_JSON_STRING payload = NULL;
+    int payloadSize = 0;
+
+    if (NULL == client)
+    {
+        LOG_ERROR("Invalid (null) client");
+        status = EINVAL;
+    }
+    else if (NULL == bin)
+    {
+        LOG_ERROR("Invalid (null) bin");
+        status = EINVAL;
+    }
+    else if (NULL == name)
+    {
+        LOG_ERROR("Invalid (null) name");
+        status = EINVAL;
+    }
+    else if (NULL == (module = (MANAGEMENT_MODULE*)calloc(1, sizeof(MANAGEMENT_MODULE))))
```

Comment:
>else if (NULL == (module = (MANAGEMENT_MODULE*)calloc(1, sizeof(MANAGEMENT_MODULE)))) [](http://example.com/codeflow?start=4&length=85)

Where is this memory freed?

Reply:
Make sure caller does

Reply:
It is freed by the caller. Ensured that the caller properly unloads modules on failure and correctly cleans up any allocated memory.